File: robolog/output-webview/src/decoder.ts
// Based on the python decoder example.
// Used: https://extendsclass.com/python-to-javascript.html

import parseISO from "date-fns/parseISO";

export interface IMessage {
    readonly message_type: string;
    readonly decoded: any;
}

function simple_decode(decoder, message) {
    return JSON.parse(message);
}

function decode_time(decoder, time) {
    const parsedTime = parseISO(time);
    // Note: doing toString() in an utc time (which ends with +00:00)
    // will convert it to the local timezone.
    return { time: parsedTime.toString() };
}

function decode_memo(decoder, message) {
    var memo_id, memo_value;
    const split = splitInChar(message, ":");
    if (split) {
        [memo_id, memo_value] = split;
        try {
            memo_value = JSON.parse(memo_value);
        } catch (error) {
            console.log("Error parsing json: " + memo_value);
            console.log(error);
            return null;
        }
        decoder.memo[memo_id] = memo_value;
    }

    return null;
}

function _decodeOid(decoder, oid) {
    const ret = decoder.memo[oid];
    if (ret === undefined) {
        return `<ref not found: ${oid}>`;
    }
    return ret;
}
function _decodeFloat(decoder, msg) {
    return parseFloat(msg);
}
function _decodeInt(decoder, msg) {
    return parseInt(msg);
}

function _decodeStr(decoder, msg) {
    return msg;
}

function _decode(message_definiton) {
    const names = [];
    const nameToDecode = new Map();
    for (let s of message_definiton.split(",")) {
        s = s.trim();
        const i = s.indexOf(":");
        let decode = "oid";
        if (i != -1) {
            [s, decode] = s.split(":", 2);
        }
        names.push(s);
        if (decode === "oid") {
            nameToDecode.set(s, _decodeOid);
        } else if (decode === "int") {
            nameToDecode.set(s, _decodeInt);
        } else if (decode === "float") {
            nameToDecode.set(s, _decodeFloat);
        } else if (decode === "str") {
            nameToDecode.set(s, _decodeStr);
        } else {
            throw new Error("Unexpected: " + decode);
        }
    }

    function _decImpl(decoder, message) {
        const splitted = message.split("|", names.length);
        const ret = {};
        for (let index = 0; index < splitted.length; index++) {
            const s = splitted[index];
            const name = names[index];
            ret[name] = nameToDecode.get(name)(decoder, s);
        }
        // console.log("decoded", ret);
        return ret;
    }
    return _decImpl;
}

const id_decode = _decode("part:int, id:str");

const _MESSAGE_TYPE_INFO = {
    "V": _decode("version:str"),
    "ID": id_decode,
    "I": simple_decode,
    "T": decode_time,
    "M": decode_memo,
    "SR": _decode("name:oid, time_delta_in_seconds:float"),
    "ER": _decode("status:oid, time_delta_in_seconds:float"),
    "ST": _decode("name:oid, libname:oid, source:oid, lineno:int, time_delta_in_seconds:float"),
    "ET": _decode("status:oid, message:oid, time_delta_in_seconds:float"),
    "SE": _decode("name:oid, libname:oid, type:oid, doc:oid, source:oid, lineno:int, time_delta_in_seconds:float"),
    "EE": _decode("type:oid, status:oid, time_delta_in_seconds:float"),
    "EA": _decode("name:oid, type:oid, value:oid"),
    "AS": _decode("source:oid, lineno:int, target:oid, type:oid, value:oid, time_delta_in_seconds:float"),
    "L": _decode("level:str, message:oid, source:oid, lineno:int, time_delta_in_seconds:float"),
    "LH": _decode("level:str, message:oid, source:oid, lineno:int, time_delta_in_seconds:float"),
    "TG": _decode("tag:oid"),
    "S": _decode("start_time_delta:float"),
    "STB": _decode("message:oid, time_delta_in_seconds:float"),
    "RTB": _decode("message:oid, time_delta_in_seconds:float"),
    "TBE": _decode("source:oid, lineno:int, method:oid, line_content:oid"),
    "TBV": _decode("name:oid, type:oid, value:oid"),
    "ETB": _decode("time_delta_in_seconds:float"),
    "YS": _decode("source:oid, lineno:int, type:oid, value:oid, time_delta_in_seconds:float"),
    "YR": _decode("name:oid, libname:oid, source:oid, lineno:int, time_delta_in_seconds:float"),
};

_MESSAGE_TYPE_INFO["RR"] = _MESSAGE_TYPE_INFO["SR"];
_MESSAGE_TYPE_INFO["RT"] = _MESSAGE_TYPE_INFO["ST"];
_MESSAGE_TYPE_INFO["RE"] = _MESSAGE_TYPE_INFO["SE"];
_MESSAGE_TYPE_INFO["RYR"] = _MESSAGE_TYPE_INFO["YR"];

export class Decoder {
    memo;

    constructor() {
        this.memo = {};
    }

    decode_message_type(message_type, message) {
        var handler;
        handler = _MESSAGE_TYPE_INFO[message_type];
        return handler(this, message);
    }
}

function splitInChar(line: string, char: string) {
    const i = line.indexOf(char);
    if (i > 0) {
        const message_type = line.substring(0, i);
        const message = line.substring(i + 1);
        return [message_type, message];
    }
    return undefined;
}

export function* iter_decoded_log_format(stream: string, decoder: Decoder) {
    var decoded, message, message_type;

    for (let line of stream.split(/\r?\n/)) {
        line = line.trim();

        if (line) {
            try {
                const split = splitInChar(line, " ");
                if (split) {
                    [message_type, message] = split;
                    decoded = decoder.decode_message_type(message_type, message);

                    if (decoded) {
                        const m: IMessage = { "message_type": message_type, "decoded": decoded };
                        yield m;
                    }
                }
            } catch (err) {
                console.log("Unable to decode message: " + line);
                console.log(err);
            }
        }
    }
}

File: log/src/robocorp/log/_decoder.py
import datetime
import json
from logging import getLogger
from typing import Any, Callable, Dict, Iterator, List, Optional, Tuple

from .protocols import IReadLines

# Whenever the decoding changes we should bump up this version.
DOC_VERSION = "0.0.2"

# name, libname, source, docstring, lineno
Location = Tuple[str, str, str, str, int]

_LOGGER = getLogger(__name__)


class Decoder:
    def __init__(self) -> None:
        self.memo: Dict[str, str] = {}
        self.location_memo: Dict[str, Location] = {}

    def decode_message_type(self, message_type: str, message: str) -> Optional[dict]:
        handler = _MESSAGE_TYPE_INFO[message_type]
        ret = {"message_type": message_type}
        try:
            r = handler(self, message)
            if not r:
                if message_type in ("M", "P"):
                    return None
                raise RuntimeError(
                    f"No return when decoding: {message_type} - {message}"
                )
                if not isinstance(r, dict):
                    ret[
                        "error"
                    ] = f"Expected dict return when decoding: {message_type} - {message}. Found: {ret}"

            ret.update(r)
        except Exception as e:
            ret["error"] = f"Error decoding: {message_type}: {e}"
        return ret


def _decode_oid(decoder: Decoder, oid: str) -> str:
    return decoder.memo[oid]


def _decode_float(decoder: Decoder, msg: str) -> float:
    return float(msg)


def _decode_int(decoder: Decoder, msg: str) -> int:
    return int(msg)


def _decode_str(decoder: Decoder, msg: str) -> str:
    return msg


def _decode_json(decoder: Decoder, msg: str) -> str:
    return json.loads(msg)


def _decode_dateisoformat(decoder: Decoder, msg: str) -> str:
    d: datetime.datetime = datetime.datetime.fromisoformat(msg)
    # The internal time is in utc, so, we need to decode it to the current timezone.
    d = d.astimezone()
    return d.isoformat(timespec="milliseconds")


def _decode(message_definition: str) -> Callable[[Decoder, str], Any]:
    if message_definition == "memorize":
        return decode_memo
    elif message_definition == "memorize_path":
        return decode_path_location

    names: List[str] = []
    name_to_decode: dict = {}
    for s in message_definition.split(","):
        s = s.strip()
        i = s.find(":")
        if i != -1:
            s, decode = s.split(":", 1)
        else:
            raise AssertionError(f"Unexpected definition: {message_definition}")

        names.append(s)
        if decode == "oid":
            name_to_decode[s] = _decode_oid

        elif decode == "int":
            name_to_decode[s] = _decode_int

        elif decode == "float":
            name_to_decode[s] = _decode_float

        elif decode == "str":
            name_to_decode[s] = _decode_str

        elif decode == "json.loads":
            name_to_decode[s] = _decode_json

        elif decode == "dateisoformat":
            name_to_decode[s] = _decode_dateisoformat

        elif decode == "loc_id":
            name_to_decode[s] = "loc_id"

        elif decode == "loc_and_doc_id":
            name_to_decode[s] = "loc_and_doc_id"

        else:
            raise RuntimeError(f"Unexpected: {decode}")

    def dec_impl(decoder: Decoder, message: str):
        splitted = message.split("|", len(names) - 1)
        ret: Dict[str, Any] = {}
        for i, s in enumerate(splitted):
            name = names[i]
            try:
                dec_func = name_to_decode[name]
                if dec_func == "loc_id":
                    try:
                        info = decoder.location_memo[s]
                    except:
                        _LOGGER.critical(f"Could not find memo: {s}")
                        raise
                    ret["name"] = info[0]
                    ret["libname"] = info[1]
                    ret["source"] = info[2]
                    ret["lineno"] = info[4]

                elif dec_func == "loc_and_doc_id":
                    try:
                        info = decoder.location_memo[s]
                    except:
                        _LOGGER.critical(f"Could not find location_memo: {s}")
                        raise
                    ret["name"] = info[0]
                    ret["libname"] = info[1]
                    ret["source"] = info[2]
                    ret["doc"] = info[3]
                    ret["lineno"] = info[4]

                else:
                    ret[name] = dec_func(decoder, s)
            except:
                ret[name] = None
        return ret

    return dec_impl


def decode_memo(decoder: Decoder, message: str) -> None:
    """
    Args:
        message: something as 'a:"Start Suite"'
    """
    memo_id: str
    memo_value: str

    memo_id, memo_value = message.split(":", 1)

    # Note: while the json.loads could actually load anything, in the spec we only
    # have oid for string messages (which is why it's ok to type it as that).
    memo_value = json.loads(memo_value)
    decoder.memo[memo_id] = memo_value


def decode_path_location(decoder: Decoder, message: str) -> None:
    """
    Args:
        message: something as 'a:b|c|d|e|33'
    """
    memo_id, memo_references = message.split(":", 1)
    name_id, libname_id, source_id, doc_id, lineno = memo_references.split("|", 4)

    decoder.location_memo[memo_id] = (
        decoder.memo[name_id],
        decoder.memo[libname_id],
        decoder.memo[source_id],
        decoder.memo[doc_id],
        int(lineno),
    )


MESSAGE_TYPE_YIELD_RESUME = "YR"
MESSAGE_TYPE_YIELD_SUSPEND = "YS"
MESSAGE_TYPE_YIELD_FROM_RESUME = "YFR"
MESSAGE_TYPE_YIELD_FROM_SUSPEND = "YFS"

# Note: the spec appears in 3 places (and needs to be manually updated accordingly).
# _decoder.py (this file)
# decoder.ts
# format.md

_MESSAGE_TYPE_INFO: Dict[str, Callable[[Decoder, str], Any]] = {}


def _build_decoding():
    from robocorp.log._decoder_spec import SPEC

    for line in SPEC.splitlines(keepends=False):
        line = line.strip()
        if not line or line.startswith("#"):
            continue

        try:
            key, val = line.split(":", 1)
            key = key.strip()
            val = val.strip()
        except ValueError:
            # No ':' there, must be = msg
            key, val = line.split("=", 1)
            key = key.strip()
            val = val.strip()
            _MESSAGE_TYPE_INFO[key] = _MESSAGE_TYPE_INFO[val]
        else:
            _MESSAGE_TYPE_INFO[key] = _decode(val)


_build_decoding()


def iter_decoded_log_format(stream: IReadLines) -> Iterator[dict]:
    decoder: Decoder = Decoder()
    line: str
    message_type: str
    message: str
    decoded: Optional[dict]

    for line in stream.readlines():
        line = line.strip()
        if line:
            try:
                message_type, message = line.split(" ", 1)
            except:
                raise RuntimeError(f"Error decoding line: {line}")
            decoded = decoder.decode_message_type(message_type, message)
            if decoded:
                yield decoded
